From 0000000000000000000000000000000000000000 Mon Sep 17 00:00:00 2001
From: RLtools <rl-tools@localhost>
Date: Sun, 6 Sep 2026 00:00:00 +0000
Subject: [PATCH 2/8] [CMake] Recognize WASI hosts and test for setjmp and
 sys/resource.h

A WASI host (CMAKE_SYSTEM_NAME=WASI) is Unix-like as far as LLVM's
platform abstraction goes: the Unix implementations of Path, Process,
Program and friends apply, but wasm-ld has no version scripts.

Two things every other Unix takes for granted are optional there:
wasi-libc provides setjmp/longjmp only when built for exception-handling
based unwinding, and it provides sys/resource.h only as an opt-in
emulation that still leaves getrlimit and setrlimit unimplemented. Both
become configure-time checks (HAVE_SETJMP, HAVE_SYS_RESOURCE_H) that the
Support library can consult instead of assuming them on every Unix.
---
 llvm/cmake/config-ix.cmake                 | 6 ++++++
 llvm/cmake/modules/HandleLLVMOptions.cmake | 5 +++++
 llvm/include/llvm/Config/config.h.cmake    | 6 ++++++
 3 files changed, 17 insertions(+)

diff --git a/llvm/cmake/config-ix.cmake b/llvm/cmake/config-ix.cmake
index ed2bfa6df..6484716ca 100644
--- a/llvm/cmake/config-ix.cmake
+++ b/llvm/cmake/config-ix.cmake
@@ -56,6 +56,9 @@ else()
   check_include_file(unistd.h HAVE_UNISTD_H)
   check_include_file(sys/ioctl.h HAVE_SYS_IOCTL_H)
 endif()
+# Not part of the platform lists above: wasi-libc provides sys/resource.h only
+# as an opt-in emulation and implements no resource limits either way.
+check_include_file(sys/resource.h HAVE_SYS_RESOURCE_H)
 
 if( UNIX AND NOT (APPLE OR BEOS OR HAIKU) )
   # Used by check_symbol_exists:
@@ -381,6 +384,9 @@ check_symbol_exists(malloc_zone_statistics malloc/malloc.h
 check_symbol_exists(posix_spawn spawn.h HAVE_POSIX_SPAWN)
 check_symbol_exists(pread unistd.h HAVE_PREAD)
 check_symbol_exists(sbrk unistd.h HAVE_SBRK)
+# setjmp is optional on WebAssembly: wasi-libc only provides it when built
+# with exception-handling based unwinding.
+check_symbol_exists(setjmp setjmp.h HAVE_SETJMP)
 check_symbol_exists(strerror_r string.h HAVE_STRERROR_R)
 check_symbol_exists(strerror_s string.h HAVE_DECL_STRERROR_S)
 check_symbol_exists(setenv stdlib.h HAVE_SETENV)
diff --git a/llvm/cmake/modules/HandleLLVMOptions.cmake b/llvm/cmake/modules/HandleLLVMOptions.cmake
index 311123084..faad15fb7 100644
--- a/llvm/cmake/modules/HandleLLVMOptions.cmake
+++ b/llvm/cmake/modules/HandleLLVMOptions.cmake
@@ -242,6 +242,11 @@ elseif(FUCHSIA OR UNIX OR CYGWIN)
   else()
     set(LLVM_HAVE_LINK_VERSION_SCRIPT 1)
   endif()
+elseif(WASI)
+  # WASI is Unix-like for LLVM's purposes (the Unix/*.inc implementations
+  # apply), but wasm-ld has no version scripts.
+  set(LLVM_ON_UNIX 1)
+  set(LLVM_HAVE_LINK_VERSION_SCRIPT 0)
 elseif(CMAKE_SYSTEM_NAME STREQUAL "Generic")
   set(LLVM_ON_UNIX 0)
   set(LLVM_HAVE_LINK_VERSION_SCRIPT 0)
diff --git a/llvm/include/llvm/Config/config.h.cmake b/llvm/include/llvm/Config/config.h.cmake
index ce83de8e4..2a35be6fc 100644
--- a/llvm/include/llvm/Config/config.h.cmake
+++ b/llvm/include/llvm/Config/config.h.cmake
@@ -152,6 +152,9 @@
 /* Define to 1 if you have the `setenv' function. */
 #cmakedefine HAVE_SETENV ${HAVE_SETENV}
 
+/* Define to 1 if you have the `setjmp' function. */
+#cmakedefine HAVE_SETJMP ${HAVE_SETJMP}
+
 /* Define to 1 if you have the `sigaltstack' function. */
 #cmakedefine HAVE_SIGALTSTACK ${HAVE_SIGALTSTACK}
 
@@ -167,6 +170,9 @@
 /* Define to 1 if you have the <sys/ioctl.h> header file. */
 #cmakedefine HAVE_SYS_IOCTL_H ${HAVE_SYS_IOCTL_H}
 
+/* Define to 1 if you have the <sys/resource.h> header file. */
+#cmakedefine HAVE_SYS_RESOURCE_H ${HAVE_SYS_RESOURCE_H}
+
 /* Define to 1 if stat struct has st_mtimespec member .*/
 #cmakedefine HAVE_STRUCT_STAT_ST_MTIMESPEC_TV_NSEC ${HAVE_STRUCT_STAT_ST_MTIMESPEC_TV_NSEC}
 
